Mathlib/Data/Fintype/Pi.lean

```diff
@@ -134,3 +136,67 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
+  simp_rw [Fin.mem_piFinset_succ_iff, cons_zero, tail_cons]
+
+lemma Fin.mem_piFinset_succ_iff' {n : ℕ} {α : Fin (n + 1) → Type*} {p : (i : Fin (n + 1)) → α i}
+    {S : (i : Fin (n + 1)) → Finset (α i)} :
+    p ∈ Fintype.piFinset S ↔
+      Fin.init p ∈ Fintype.piFinset (Fin.init S) ∧ p (Fin.last n) ∈ S (Fin.last n) := by
+  simp only [Fintype.mem_piFinset, forall_fin_succ', Fin.init]
+
+lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
+    {xs : (i : Fin n) → α i.castSucc} {x : α (.last n)}
+    {Sᵢ : (i : Fin n) → Finset (α i.castSucc)} {Sₙ : Finset (α <| .last n)} :
+    Fin.snoc xs x ∈ Fintype.piFinset (Fin.snoc Sᵢ Sₙ) ↔ xs ∈ Fintype.piFinset Sᵢ ∧ x ∈ Sₙ := by
+  simp_rw [Fin.mem_piFinset_succ_iff', init_snoc, snoc_last]
+
+lemma Finset.map_piFinSuccAbove_filter_piFinset_succAbove {n : ℕ} {k : Fin (n + 1)}
+    {α : Fin (n + 1) → Type*}
+    {p : ((i : Fin n) → α (Fin.succAbove k i)) → Prop} [DecidablePred p]
+    {S : (i : Fin (n + 1)) → Finset (α i)} :
+    ((Fintype.piFinset S).filter fun r ↦ p (fun x ↦ r <| Fin.succAbove k x)).map
+      (Equiv.piFinSuccAbove α k).toEmbedding
+    = S k ×ˢ (Fintype.piFinset (fun x ↦ S <| Fin.succAbove k x)).filter p := by
+  congr
+  ext ⟨x, f⟩
+  simp? [Fin.forall_iff_succAbove
+        k] says simp only [mem_map_equiv, Equiv.piFinSuccAbove_symm_apply, mem_filter,
+      Fintype.mem_piFinset, Fin.forall_iff_succAbove k, Fin.insertNth_apply_same,
+      Fin.insertNth_apply_succAbove, mem_product]
+  tauto
+
+lemma Finset.map_piFinSuccAbove_filter_piFinset {n : ℕ} {α : Fin (n + 1) → Type*}
+    {p : ((i : Fin n) → α i.succ) → Prop} [DecidablePred p]
+    {S : (i : Fin (n + 1)) → Finset (α i)} :
+    ((Fintype.piFinset S).filter fun r ↦ p (Fin.tail r)).map
+      (Equiv.piFinSuccAbove α 0).toEmbedding
+    = S 0 ×ˢ (Fintype.piFinset (Fin.tail S)).filter p :=
+  Finset.map_piFinSuccAbove_filter_piFinset_succAbove
+
+lemma Finset.filter_piFinset_eq_map_piFinSuccAbove_symm {n : ℕ} {α : Fin (n + 1) → Type*}
+    {p : ((i : Fin n) → α i.succ) → Prop} [DecidablePred p]
+    {S : ∀ i, Finset (α i)} :
+    ((Fintype.piFinset S).filter fun r ↦ p (Fin.tail r))
+    = (S 0 ×ˢ (Fintype.piFinset (Fin.tail S)).filter p).map
+        (Equiv.piFinSuccAbove α 0).symm.toEmbedding := by
+  rw [← Finset.map_piFinSuccAbove_filter_piFinset, Finset.map_map,
+    Function.Embedding.equiv_toEmbedding_trans_symm_toEmbedding, map_refl]
+
+lemma Finset.card_filter_succ_piFinset_eq {n : ℕ} {α : Fin (n + 1) → Type*}
+    {p : ((i : Fin n) → α i.succ) → Prop} [DecidablePred p]
+    {S : (i : Fin (n + 1)) → Finset (α i)} :
+    ((Fintype.piFinset S).filter fun r ↦ p (Fin.tail r)).card
+    = (S 0).card * ((Fintype.piFinset (Fin.tail S)).filter p).card := by
+  rw [← Finset.card_map ((Equiv.piFinSuccAbove α 0).toEmbedding),
+    map_piFinSuccAbove_filter_piFinset]
+  exact Finset.card_product (S 0) ((Fintype.piFinset (Fin.tail S)).filter p)
```

Comment:
You're missing lemmas here, right? There are three lemmas about `S 0`, one lemma about `S k` and no lemma about `S (Fin.last n)`.

Reply:
Well `Fin.mem_piFinset_succ_iff'` contains a reference to `S (Fin.last n)` on line 154. 

Here are the cons-snoc analogies as I see them
`Fin.mem_piFinset_succ_iff` is the cons version where  `Fin.mem_piFinset_succ_iff'` is the snoc version
`Fin.cons_mem_piFinset_cons_iff` is the cons version where `Fin.snoc_mem_piFinset_snoc_iff`
`Finset.map_piFinSuccAbove_filter_piFinset_succAbove` relates to an internal value `k`

`Finset.map_piFinSuccAbove_filter_piFinset` is `cons`-y, but trying the analogous `snoc`-y lemma, I get an error
```
lemma Finset.map_piFinSuccAbove_filter_piFinset_init {n : ℕ} {α : Fin (n + 1) → Type*}
    {p : ((i : Fin n) → α i.castSucc) → Prop} [DecidablePred p]
    {S : (i : Fin (n + 1)) → Finset (α i)} :
    ((Fintype.piFinset S).filter fun r ↦ p (Fin.init r)).map
      (Equiv.piFinSuccAbove α (Fin.last n)).toEmbedding
    = S (Fin.last n) ×ˢ (Fintype.piFinset (Fin.init S)).filter p -- error on this line
:= by 
  sorry
```
```
type mismatch
  S (Fin.last n) ×ˢ filter p (Fintype.piFinset (Fin.init S))
has type
  Finset (α (Fin.last n) × ((i : Fin n) → α (Fin.castSucc i))) : outParam (Type u_2)
but is expected to have type
  Finset (α (Fin.last n) × ((j : Fin n) → α (Fin.succAbove (Fin.last n) j))) : Type u_2
```
I am not sure why this poses a problem here while the `cons`-y one doesn't. Perhaps there is some definitional equality for that one. I suppose I could solve it with a cast but that seems like more complicated machinery. The last three lemmas all depend on this one so I can't prove the `snoc`-y version here.

Reply:
That's because you're using `castSucc` in the finset but `succAbove (last n)` in the map. Those functions are equal but not defeq. Could you define `Equiv.piFinCastSucc` and use it here?

Reply:
[Equiv.piFinCastSucc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#Equiv.piFinCastSucc) already exists. It can't be used here because it isn't dependent the way [Equiv.piFinSuccAbove](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#Equiv.piFinSuccAbove) is.

Reply:
In fact, I see that this very [Equiv.piFinCastSucc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#Equiv.piFinCastSucc) function was something [you asked me to add](https://github.com/leanprover-community/mathlib4/pull/7898#discussion_r1423557280) for this PR back in December to match [Equiv.piFinSucc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#Equiv.piFinSucc).

Reply:
When I try to make is dependent, I get a similar error which again doesn't happen for the cons-y version
```
def Equiv.piFinCastSucc (n : ℕ) (α : Fin (n + 1) → Type u) : 
   ((i : Fin (n + 1)) → α i) ≃ (α (.last _)) × ((j : Fin n) → α (.castSucc j)) :=
  Equiv.piFinSuccAbove α (.last _)
```
```
type mismatch
  piFinSuccAbove α (Fin.last n)
has type
  ((j : Fin (n + 1)) → α j) ≃ α (Fin.last n) × ((j : Fin n) → α (Fin.succAbove (Fin.last n) j)) : Type u
but is expected to have type
  ((i : Fin (n + 1)) → α i) ≃ α (Fin.last n) × ((j : Fin n) → α (Fin.castSucc j)) : Type u
```

Reply:
I mean yeah you can't use `piFinSuccAbove` to define it (at least without some `cast` somewhere).

I agree that making `Equiv.piFinCastSucc` dependent is the correct move.

Reply:
To be clear, I am not trying to imply I do or don't think making `Equiv.piFinCastSucc` dependent is the correct move. I am just trying to point out that there are difficulties in doing so. 

I would like to express again that trying to develop `snoc` related API is not what I would like to be doing with my time. I think that probably this work would be better off being done by someone who has a better sense of why it's important, and therefore a better sense of the design considerations involved.

Reply:
> I would like to express again that trying to develop `snoc` related API is not what I would like to be doing with my time. I think that probably this work would be better off being done by someone who has a better sense of why it's important, and therefore a better sense of the design considerations involved.

The problem is that, by doing so, you are making the parallel APIs diverge, thus decreasing the quality of mathlib as a whole.

I have now opened #13567 and #13568 to add the missing prerequisites. I will probably need one further PR before yours can be merged.